object.go

```diff
@@ -140,6 +142,43 @@ func (o *Object) DecCyclicalCount() {
 	o.cyclicalCount--
 }
 
+func (o *Object) AddSeenObject(ptr *Object) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+
+	if o.seenObjects == nil {
+		o.seenObjects = make(map[*Object]bool)
+	}
+	if _, ok := o.seenObjects[ptr]; !ok {
+		o.seenObjects[ptr] = true
+	}
+}
+
+func (o *Object) AddSeenObjectMap(seen map[*Object]bool) {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+
+	if o.seenObjects == nil {
+		o.seenObjects = seen
+	} else {
+		for ptr := range seen {
+			o.seenObjects[ptr] = true
+		}
+	}
```

Comment:
it's more idiomatic go to drop this else statement in favor of a return in the if block:
```
if o.seenObjects == nil {
		o.seenObjects = seen
return
	}

for loop
```